capacity4more/modules/c4m/features_og/c4m_features_og_events/c4m_features_og_events.module

```diff
@@ -219,38 +219,110 @@ function c4m_features_og_events_entity_info_alter(&$entity_info) {
  * 1 Event for anonymous user, 3 for logged in.
  *
  * @return string
- *  HTML of rendered upcoming events.
+ *  HTML of rendered upcoming events, FALSE if there's no events.
  */
 function c4m_features_og_events_upcoming_events() {
   global $user;
   $events = '';
+  $filter = '';
 
   // Show all events starting from midnight onwards.
   $tomorrow = new DateTime('midnight tomorrow');
   $date = $tomorrow->format('Y-m-d H:i:s');
 
-  // User's group when user is logged in, Public groups when not.
-  $groups = $user->uid ? og_get_groups_by_user() : c4m_og_get_groups_by_access_type('public');
+  // Get user's interests when user is logged in.
+  // Get content filter when user is logged in.
+  // Get user's group when user is logged in, Public groups when not.
   // 3 Events when user is logged in, 1 when not.
-  $range = $user->uid ? 3 : 1;
+  if ($user->uid) {
+    // Check if content filter is applied.
+    $user_wrapper = entity_metadata_wrapper('user', $user->uid);
+    $interests = $user_wrapper->c4m_related_topic->raw();
+    $filter = !empty($_GET['filter']) ? $_GET['filter'] : '';
+    $groups = array();
+    $range = 3;
+  }
+  else {
+    $groups = c4m_og_get_groups_by_access_type('public');
```

Comment:
It's not only the events of public groups a user has access to.

It can be that the user has an email address domain that gives him access to all restricted groups who granted access to this domain.

### Solution
Get 3 events that:

* Startdate >= midnight tomorrow
* Filtered by the homepage filter:
  * Any
  * By interests (topics)
  * Within my groups
* Show only events I have access to (use the node_access system).